source/extensions/common/aws/utility.cc

```diff
@@ -234,19 +234,45 @@ Utility::joinCanonicalHeaderNames(const std::map<std::string, std::string>& cano
 }
 
 std::string Utility::getSTSEndpoint(absl::string_view region) {
-  if (region == "cn-northwest-1" || region == "cn-north-1") {
-    return fmt::format("sts.{}.amazonaws.com.cn", region);
+
+  // Handle the scenario where we are using SigV4A and a region set has been provided
+  // SigV4A region sets are comma separated and may include wildcard
+
+  std::string single_region;
+
+  // If we contain a comma or asterisk it looks like a region set
+  if (absl::StrContains(region, ",") || (absl::StrContains(region, "*"))) {
+    // Use the first element from a region set if we have multiple regions specified
+    std::vector<std::string> region_v = absl::StrSplit(region, ',');
+    // If we still have a * in the first element, then send them to us-east-1 fips or global
+    // endpoint
+    if (absl::StrContains(region_v[0], '*')) {
+#ifdef ENVOY_SSL_FIPS
+      return "sts-fips.us-east-1.amazonaws.com";
+#else
+      return "sts.amazonaws.com";
+#endif
+    }
+    single_region = region_v[0];
+  } else {
+    // Otherwise it's a standard region, so use that
+    single_region = region;
   }
+
+  if (single_region == "cn-northwest-1" || single_region == "cn-north-1") {
+    return fmt::format("sts.{}.amazonaws.com.cn", single_region);
```

Comment:
Is sigv4a supported in China partition?

If it is supported, then we might end up with global sts endpoint (aws commercial partition) when some one specify it as `cn-*`.

If Sigv4a is not support yet in China partition, then we can mention a TODO.

Reply:
This is actually a harder problem to solve and linked to the original decision here https://github.com/envoyproxy/envoy/pull/31336#discussion_r1425738142 

We don't capture partition details anywhere, so there is no correct way to generate a generic sts endpoint - ie it could be cn, or govcloud, or any other partition. This would always occur in the common and worst case, where region is set to '*'

The problem is compounded because we overload region and a sigv4a region set - so you can't specify a different region (so that webidentity provider goes to the right place) and a region set (used only in signing)

I think the only way to correctly solve this is to add region_set as an optional proto as well as an optional partition proto.

Reply:
another option would be to add a proto to explicitly set the sts - something like explicit_sts_host: <hostname> - this would allow you to override automatic sts generation if in a non standard partition, or if you are making it generic for sigv4a

Reply:
May be let's add this caveat in docs saying the global region set defaults to sts in us-east-1, which should be ok for AWS commercial partition (majority usecase). Also let's add a debug log on what was the input and what was the sts region that got derived.

In docs we can add a note saying, in China or GovCloud partition please explicitly provide the region set instead of using wild card, which should be ok since those are few in number.